packages/foam-core/src/model/uri.ts

```diff
@@ -0,0 +1,508 @@
+// Some code coming from
+// https://github.com/microsoft/vscode/tree/master/src/vs/base/common/uri.ts
+
+import * as paths from 'path';
+import GithubSlugger from 'github-slugger';
+import { hash } from '../utils/core';
+import { statSync } from 'fs';
+import { CharCode } from '../common/charCode';
+import { isWindows } from '../common/platform';
+
+/**
+ * Uniform Resource Identifier (URI) http://tools.ietf.org/html/rfc3986.
+ * This class is a simple parser which creates the basic component parts
+ * (http://tools.ietf.org/html/rfc3986#section-3) with minimal validation
+ * and encoding.
+ *
+ * ```txt
+ *       foo://example.com:8042/over/there?name=ferret#nose
+ *       \_/   \______________/\_________/ \_________/ \__/
+ *        |           |            |            |        |
+ *     scheme     authority       path        query   fragment
+ *        |   _____________________|__
+ *       / \ /                        \
+ *       urn:example:animal:ferret:nose
+ * ```
+ */
+export interface URI {
+  scheme: string;
+  authority: string;
+  path: string;
+  query: string;
+  fragment: string;
+}
+
+const { posix } = paths;
+const _empty = '';
+const _slash = '/';
+const _regexp = /^(([^:/?#]+?):)?(\/\/([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?/;
+
+export const create = (
```

Comment:
Consider refactoring similar to proposed here https://github.com/foambubble/foam/pull/532/files#r597580366